compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ),
                 Mismatch::Fixed(s) => (s.into(), s.into(), None),
             };
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+            let looks_similar = |e: ExpectedFound<Ty<'_>>| {
+                // We're only interested in adts
+                if let (Some(e), Some(f)) = (e.expected.ty_adt_def(), e.found.ty_adt_def()) {
+                    // Only compare the last parts of the path.
+                    // `whatever::Foo` is pretty similar to `blah::Foo`
+                    let e_path = self.tcx.def_path(e.did()).data;
+                    let f_path = self.tcx.def_path(f.did()).data;
+                    if let (Some(e), Some(f)) = (e_path.last(), f_path.last()) {
+                        return e.data == f.data;
+                    }
+                }
+                false
+            };
+
+            match terr {
+                // If two types mismatch but have similar names, mention that specifically.
+                TypeError::Sorts(values) if looks_similar(values) => {
+                    let found_adt = values.found.ty_adt_def().unwrap();
+                    let expected_adt = values.expected.ty_adt_def().unwrap();
+
+                    let found_name = values.found.sort_string(self.tcx);
+                    let expected_name = values.expected.sort_string(self.tcx);
+
+                    diag.note(format!("{found_name} and {expected_name} have similar names, but are actually distinct types"));
+
+                    for (adt, name) in [(found_adt, found_name), (expected_adt, expected_name)] {
+                        let defid = adt.did();
+                        let def_span = self.tcx.def_span(defid);
+
+                        let msg = if defid.is_local() {
+                            format!("{name} is defined in the current crate.")
+                        } else if self.tcx.all_diagnostic_items(()).id_to_name.get(&defid).is_some()
+                        {
+                            // if it's a diagnostic item, it's definitely defined in std/core/alloc
+                            // otherwise might be, might not be.
+                            format!("{name} is defined in the standard library.")
+                        } else {
+                            let crate_name = self.tcx.crate_name(defid.krate);
+                            format!("{name} is defined in crate `{crate_name}`.")
```

Comment:
having this inconsistency between libstd types that are diagnostic items and types that aren't seems unfortunate. I think we should either switch on `crate_name() == std/core/alloc` or use `crate {name}` consistently. My preference is for the latter; if two types with the same name are defined in core and std, this message will not be particularly helpful.

Reply:
Let's do the latter :)